core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -107,10 +107,34 @@ object Partition extends KafkaMetricsGroup {
   def apply(topicPartition: TopicPartition,
             time: Time,
             replicaManager: ReplicaManager): Partition = {
-    val zkIsrBackingStore = new ZkPartitionStateStore(
-      topicPartition,
-      replicaManager.zkClient,
-      replicaManager)
+    val backingStore: PartitionStateStore = if (replicaManager.zkClient.isDefined) {
```

Comment:
It seems like implementing the interface, but then throwing `IllegalStateException` when we try to call most of the methods, is kind of ugly here.

What I think would be good here is to have three implementations of the interface: the existing ZkPartitionStateStore one, one that uses AlterIsr, and another one that uses either ZK or AlterIsr based on the IBP.

Reply:
This interface seems like kind of a grab-bag to me.  I think there's really 2 separate things going on here.  There's a `"TopicConfigRetriever"` (the `fetchTopicConfig(): Properties` method) and an `"IsrChangerViaZooKeeper"` (the `{shrink,expand}Isr(controllerEpoch: Int, leaderAndIsr: LeaderAndIsr): Option[Int]` methods).  The topic config retriever is used in **all** configurations (whenever a new Log has to be created), but the ZK Isr changer is **only** used when ZooKeeper is the AlterIsr communication mechanism.  What do you think about explicitly separating these two things out and having `Partition` take a mandatory topic config retriever and an `Option[]` for the ZK Isr changer?  When the partition instance gets constructed it can do the check on the ZK Isr changer Option to make sure it actually has something when `useAlterisr` is false.

Reply:
@mumrah will take a look at this in KC-1264.  I'll close this comment thread for now